File: Examples and Theory/Graphs/DSU.py
'''

parent = [None]*10000


# this is  BASIC OVERVIEW OF DFS
def make(a):
    parent[a] = a

# recrursive
def find(v):
    if v == parent[v]: return v
    return find(parent[v])

# iterative
def find_(v):
    while v != parent[v]:
        v = parent[v]
    return v

def union(a, b):
    a = find(a)
    b = find(b)
    if a!=b: parent[b] = a

'''


# BUT FURTHUR OPTIMIZATION IS POSSIBLE
parent = [i for i in range(20)]
size = [1]*20

def find(node):
    if parent[node] == node: 
        return node

    parent[node] = find(parent[node])
    return parent[node]


def union(a, b):

    print(f'-> Before path compression : parent of {a}: {parent[a]}, parent of {b}: {parent[b]}')

    a = find(a)
    b = find(b)


    if a!=b:
        if size[a] < size[b]:
            a, b = b, a

        # we are sure that component of A has more nodes than B
        parent[b] = a
        size[a] += size[b]
    print(f'-> After  path compression : parents of _: {parent[a]}, parent of _: {parent[b]}')


def DSU_workings(u,v):

    print(f'\n----------Adding {u} and {v} to DSU----------\n\n')
    union(u,v)

    print(f'\n\nSo parents array : {parent}')
    print(f'corresponding    : {[i for i in range(20)]}')
    print(f'corresponding    : {size}')




connect_nodes = [
    (0,10),
    (11,9),
    (0,9),
    (18,19),
    (1,2),
    ]

for u,v in connect_nodes: 
    DSU_workings(u,v)

File: Examples and Theory/Graphs/DSU.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long


// global stuff will be initialized fine
ll n,m,k,q,x,y,z;
ll u,v;
const ll arr_template_size = 1e6 + 8432;

// global int   arrays are initialized to 0
ll a[arr_template_size];
ll b[arr_template_size];
ll c[arr_template_size];

string s,t;
ll ans;




// THIS IS THE BASIC OVERVIEW OF ALL THE IMPRTANT STUFF

ll parent[arr_template_size];

void make(ll v){
    parent[v]=v;
}

ll find(ll v){
    if (parent[v]==v)return v;
    return find(parent[v]);
}

ll union1(ll a, ll b){
    a = find(a); b = find(b);
    if (a!=b)parent[b]=a;
}

// create a union function in a DSU that uses union by rank




// BUT FURTHUR OPTIMIZATION IS POSSIBLE

// union by rank ----> when we union two sets we union based on rank
// union by size ----> when we union two sets we union based on size
// path compression ----> this is the inverse ackerman function Time complexity


ll parent_[arr_template_size];
ll size_[arr_template_size];
ll rank[arr_template_size];

void make_optimized(ll v){
    parent[v]=v;
    size_[v]=1;
}


ll find_optimized(ll v){
    if (parent[v]==v)return v;

    // ------ PATH COMPRESSION -------
    // otpimized , now all the nodes along 
    // the path will point to the same parent
    // this decreses the entire path to a depth of 1 !!!!!
    return parent_[v] = find_optimized(parent_[v]);
}


ll union_by_size(ll a, ll b){
    a = find_optimized(a); b = find_optimized(b);
    if (a!=b){

        // OPTIMIZED. now we can be sure that we'll 
        // only join the smaller tree to bigger tree
        if (size_[a]<size_[b])swap(a,b); 

        parent_[b]=a;
        size_[a] = size_[b] = size_[a]+size_[b];
    }
}
                


int main(){
    cin >> n >> k;
    for(int i=0; i<n; i++)make(i);

    while(k--){
        cin >> u >> v;
        union1(u,v);
        union_by_size(u,v);
    }

    ll connected_ct = 0;
    for (int i=0; i<n; i++) 
        if (find_optimized(i) == i)
            connected_ct++;

    cout << connected_ct << endl;
}